### PR TITLE
feat(operator): merkle root check

### PR DESCRIPTION
**Pull request by uri-99**
This PR adds the merkle root check on the operator side.
Right after downloading the batch from S3, the operator will check if the merkle root of the downloaded file is the same as the expected/submitted merkle root.

To test this you can
1) run locally: anvil, aggregator, operator, batcher.
2) send proofs, normally, they should pass
3) manually go to the code and hardcode some bytecode or expected merkle root, it shoulndt pass.

### CODE REVIEW

File: operator/merkle_tree/lib/src/lib.rs [hunk 1]
```diff
@@ -1,21 +1,40 @@
...
 ) -> bool {
-
-    let batch: Vec<VerificationData> = serde_json::from_slice(&batch_bytes[..batch_len]).unwrap();
-    let batch_merkle_tree: MerkleTree<VerificationBatch> = MerkleTree::build(&batch);
-    let batch_merkle_root = hex::encode(batch_merkle_tree.root);
+    if batch_ptr.is_null() || batch_len == 0 {
+        println!("batch_ptr == null || batch_len == 0 {}", batch_len);
+        return false;
+    }
+    println!("batch_len: {}", batch_len);
```
Comment (uri-99) [inline operator/merkle_tree/lib/src/lib.rs]:
```suggestion

```

File: operator/merkle_tree/lib/src/lib.rs [hunk 2]
```diff
@@ -1,21 +1,40 @@
...
+    
+    let batch_bytes = unsafe { std::slice::from_raw_parts(batch_ptr, batch_len) };
+    println!("after unsafe: {}", batch_bytes.clone().len());
```
Comment (uri-99) [inline operator/merkle_tree/lib/src/lib.rs]:
```suggestion

```

File: operator/pkg/operator.go [hunk 3]
```diff
@@ -181,6 +181,7 @@ func (o *Operator) ProcessNewBatchLog(newBatchLog *servicemanager.ContractAligne
 	verificationDataBatchLen := len(verificationDataBatch)
 	results := make(chan bool, verificationDataBatchLen)
 	var wg sync.WaitGroup
+	fmt.Println("in some place");
```
Comment (uri-99) [inline operator/pkg/operator.go]:
```suggestion
```

File: operator/pkg/operator.go [hunk 4]
```diff
@@ -340,6 +341,8 @@ func (o *Operator) verify(verificationData VerificationData, results chan bool)
 		o.Logger.Infof("Halo2-KZG proof verification result: %t", verificationResult)
 		results <- verificationResult
 	case common.Risc0:
+		fmt.Println("in common.Risc0:");
+
```
Comment (uri-99) [inline operator/pkg/operator.go]:
```suggestion
```

File: operator/risc_zero/lib/src/lib.rs [hunk 5]
```diff
@@ -7,20 +7,28 @@ pub extern "C" fn verify_risc_zero_receipt_ffi(
     image_id: *const u8,
     image_id_len: u32,
 ) -> bool {
+    println!("in verify_risc_zero_receipt_ffi");
     if receipt_bytes.is_null() || image_id.is_null() {
         return false;
     }
 
+    println!("in verify_risc_zero_receipt_ffi after check");
```
Comment (uri-99) [inline operator/risc_zero/lib/src/lib.rs]:
```suggestion
```

File: operator/risc_zero/lib/src/lib.rs [hunk 6]
```diff
@@ -7,20 +7,28 @@ pub extern "C" fn verify_risc_zero_receipt_ffi(
...
+
     let receipt_bytes = unsafe { std::slice::from_raw_parts(receipt_bytes, receipt_len as usize) };
+    println!("in verify_risc_zero_receipt_ffi after unsafe 1");
```
Comment (uri-99) [inline operator/risc_zero/lib/src/lib.rs]:
```suggestion
```

File: operator/risc_zero/lib/src/lib.rs [hunk 7]
```diff
@@ -7,20 +7,28 @@ pub extern "C" fn verify_risc_zero_receipt_ffi(
...
 
     let image_id = unsafe { std::slice::from_raw_parts(image_id, image_id_len as usize) };
+    println!("in verify_risc_zero_receipt_ffi after unsafe 2");
```
Comment (uri-99) [inline operator/risc_zero/lib/src/lib.rs]:
```suggestion
```

File: operator/risc_zero/lib/src/lib.rs [hunk 8]
```diff
@@ -7,20 +7,28 @@ pub extern "C" fn verify_risc_zero_receipt_ffi(
...
 
     let mut image_id_array = [0u8; 32];
     image_id_array.copy_from_slice(image_id);
 
     if let Ok(receipt) = bincode::deserialize::<Receipt>(receipt_bytes) {
+        println!("in verify_risc_zero_receipt_ffi returning");
```
Comment (uri-99) [inline operator/risc_zero/lib/src/lib.rs]:
```suggestion
```

File: operator/risc_zero/lib/src/lib.rs [hunk 9]
```diff
@@ -7,20 +7,28 @@ pub extern "C" fn verify_risc_zero_receipt_ffi(
...
         return receipt.verify(image_id_array).is_ok();
     }
+    println!("in verify_risc_zero_receipt_ffi returning");
+
```
Comment (uri-99) [inline operator/risc_zero/lib/src/lib.rs]:
```suggestion
```

File: operator/risc_zero/risc_zero.go [hunk 10]
```diff
@@ -7,10 +7,16 @@ package risc_zero
...
 */
 import "C"
-import "unsafe"
+import (
+	"fmt"
+	"unsafe"
+)
 
 func VerifyRiscZeroReceipt(receiptBuffer []byte, receiptLen uint32, imageIdBuffer []byte, imageIdLen uint32) bool {
+	fmt.Println("in verify_risc_zero_receipt_ffi 1")
```
Comment (uri-99) [inline operator/risc_zero/risc_zero.go]:
```suggestion
```

File: operator/risc_zero/risc_zero.go [hunk 11]
```diff
@@ -7,10 +7,16 @@ package risc_zero
...
 	receiptPtr := (*C.uchar)(unsafe.Pointer(&receiptBuffer[0]))
+	fmt.Println("in verify_risc_zero_receipt_ffi 2")
```
Comment (uri-99) [inline operator/risc_zero/risc_zero.go]:
```suggestion
```

File: operator/risc_zero/risc_zero.go [hunk 12]
```diff
@@ -7,10 +7,16 @@ package risc_zero
...
 	imageIdPtr := (*C.uchar)(unsafe.Pointer(&imageIdBuffer[0]))
+	fmt.Println("in verify_risc_zero_receipt_ffi 3")
```
Comment (uri-99) [inline operator/risc_zero/risc_zero.go]:
```suggestion
```

**Review by uri-99 — COMMENTED** [review]
WIP

File: operator/merkle_tree/lib/Cargo.toml [hunk 13]
```diff
@@ -10,7 +10,7 @@ bincode = "1.3.3"
 lambdaworks-crypto = { version = "0.7.0", features = ["serde"] }
 hex = "0.4.3"
 sha3 = "0.10.8"
-batcher = {path = "../../../batcher"}
+aligned-sdk = {path = "../../../batcher/aligned-sdk"}
```
Comment (taturosati) [inline operator/merkle_tree/lib/Cargo.toml]: Should probably get from repo

File: operator/merkle_tree/lib/src/lib.rs [hunk 14]
```diff
@@ -1,21 +1,38 @@
...
+
+    let batch_data_comm: Vec<VerificationDataCommitment> = batch.into_iter()
+                                                            .map(VerificationDataCommitment::from)
+                                                            .collect();
+
+    let computed_batch_merkle_tree: MerkleTree<VerificationCommitmentBatch> = MerkleTree::build(&batch_data_comm);
+    let computed_batch_merkle_root = hex::encode(computed_batch_merkle_tree.root);
     let received_merkle_root = hex::encode(merkle_root);
-    return batch_merkle_root == received_merkle_root;
+    println!("returning after merkle_check: {}", computed_batch_merkle_root == received_merkle_root);
```
Comment (taturosati) [inline operator/merkle_tree/lib/src/lib.rs]: remove
```suggestion
```

File: operator/risc_zero/lib/src/lib.rs [hunk 15]
```diff
@@ -7,10 +7,12 @@ pub extern "C" fn verify_risc_zero_receipt_ffi(
     image_id: *const u8,
     image_id_len: u32,
 ) -> bool {
+    println!("in verify_risc_zero_receipt_ffi");
```
Comment (taturosati) [inline operator/risc_zero/lib/src/lib.rs]: remove
```suggestion
```

**Review by taturosati — CHANGES_REQUESTED** [review]

File: operator/merkle_tree/lib/src/lib.rs [hunk 16]
```diff
@@ -1,21 +1,37 @@
...
+    batch_ptr: *const u8,
     batch_len: usize,
     merkle_root: &[u8; 32]
 ) -> bool {
-
-    let batch: Vec<VerificationData> = serde_json::from_slice(&batch_bytes[..batch_len]).unwrap();
-    let batch_merkle_tree: MerkleTree<VerificationBatch> = MerkleTree::build(&batch);
-    let batch_merkle_root = hex::encode(batch_merkle_tree.root);
+    if batch_ptr.is_null() || batch_len == 0 {
+        println!("batch_ptr == null || batch_len == 0 {}", batch_len);
```
Comment (entropidelic) [inline operator/merkle_tree/lib/src/lib.rs]: remove the debug prints

File: operator/merkle_tree/lib/src/lib.rs [hunk 17]
```diff
@@ -1,21 +1,37 @@
...
+        return false;
+    }
+    
+    let batch_bytes = unsafe { std::slice::from_raw_parts(batch_ptr, batch_len) };
+
+    let batch = match serde_json::from_slice::<Vec<VerificationData>>(batch_bytes) {
+        Ok(batch) => batch,
+        Err(e) => {
+            println!("Error: {:?}", e);
```
Comment (entropidelic) [inline operator/merkle_tree/lib/src/lib.rs]: remove

**Review by entropidelic — APPROVED** [review]

File: Makefile
Comment (NicolasRampoldi) [inline Makefile:1]: I think we should remove the changes on repetitions, also `make batcher_send_groth16_burst` works on main.

**Review by NicolasRampoldi — CHANGES_REQUESTED** [review]

File: operator/pkg/s3.go [hunk 18]
```diff
@@ -35,14 +37,22 @@ func (o *Operator) getBatchFromS3(proofUrl string) ([]VerificationData, error) {
...
 	if err != nil {
 		return nil, err
 	}
 
+	// Checks if downloaded merkle root is the same as the expected one
+	o.Logger.Infof("Verifying batch merkle tree...")
+	merkle_root_check := merkle_tree.VerifyMerkleTreeBatch(batchBytes, uint(len(batchBytes)), expectedMerkleRoot)
+	if !merkle_root_check {
+		return nil, fmt.Errorf("merkle root check failed")
+	}
```
Comment (NicolasRampoldi) [inline operator/pkg/s3.go:50]: We shouldn't return an error, instead we should log a warning

Comment (uri-99) [reply operator/pkg/s3.go:50]: We also do `return fmt.Errorf("invalid proof")` on line 206 of operator.go

File: operator/merkle_tree/lib/src/lib.rs [hunk 19]
```diff
@@ -1,21 +1,35 @@
...
+    let batch_data_comm: Vec<VerificationDataCommitment> = batch.into_iter()
+                                                            .map(VerificationDataCommitment::from)
+                                                            .collect();
+
+    let computed_batch_merkle_tree: MerkleTree<VerificationCommitmentBatch> = MerkleTree::build(&batch_data_comm);
+    let computed_batch_merkle_root = hex::encode(computed_batch_merkle_tree.root);
     let received_merkle_root = hex::encode(merkle_root);
-    return batch_merkle_root == received_merkle_root;
+
+    return computed_batch_merkle_root == received_merkle_root;
```
Comment (taturosati) [inline operator/merkle_tree/lib/src/lib.rs]: I think you can check this without unnecessarily encoding to hex

File: operator/risc_zero/lib/src/lib.rs [hunk 20]
```diff
@@ -13,6 +13,7 @@ pub extern "C" fn verify_risc_zero_receipt_ffi(
         return false;
     }
 
+
```
Comment (taturosati) [inline operator/risc_zero/lib/src/lib.rs]:
```suggestion
```

File: operator/merkle_tree/lib/src/lib.rs (hunk 19 above)
Comment (uri-99) [reply operator/merkle_tree/lib/src/lib.rs]: It doesn't compile without `hex::encode()`

Comment (taturosati) [reply operator/merkle_tree/lib/src/lib.rs]:
```rust
let computed_batch_merkle_tree: MerkleTree<VerificationCommitmentBatch> = MerkleTree::build(&batch_data_comm);

return computed_batch_merkle_tree.root == *merkle_root;
```

**Review by taturosati — APPROVED** [review]

**Review by NicolasRampoldi — APPROVED** [review]

**Pull request merged**